scarches/models/scgen/vaearith_model.py

```diff
@@ -0,0 +1,344 @@
+    """ Adapted from
+        Title: scvi-tools
+        Authors: Romain Lopez <romain_lopez@gmail.com>,
+                 Adam Gayoso <adamgayoso@berkeley.edu>,
+                 Galen Xing <gx2113@columbia.edu>
+        Date: 14.12.2020
+        Code version: 0.8.0-beta.0
+        Availability: https://github.com/YosefLab/scvi-tools
+        Link to the used code:
+        https://github.com/YosefLab/scvi-tools/blob/0.8.0-beta.0/scvi/core/models/base.py
+    """
+    def _get_user_attributes(self):
+        # returns all the self attributes defined in a model class, eg, self.is_trained_
+        attributes = inspect.getmembers(self, lambda a: not (inspect.isroutine(a)))
+        attributes = [
+            a for a in attributes if not (a[0].startswith("__") and a[0].endswith("__"))
+        ]
+        attributes = [a for a in attributes if not a[0].startswith("_abc_")]
+        return attributes
+
+    def _get_public_attributes(self):
+        public_attributes = self._get_user_attributes()
+        public_attributes = {a[0]: a[1] for a in public_attributes if a[0][-1] == "_"}
+        return public_attributes
+
+    def save(
+        self,
+        dir_path: str,
+        overwrite: bool = False,
+        save_anndata: bool = False,
+        **anndata_write_kwargs,
+    ):
+        """Save the state of the model.
+           Neither the trainer optimizer state nor the trainer history are saved.
+           Parameters
+           ----------
+           dir_path
+                Path to a directory.
+           overwrite
+                Overwrite existing data or not. If `False` and directory
+                already exists at `dir_path`, error will be raised.
+           save_anndata
+                If True, also saves the anndata
+           anndata_write_kwargs
+                Kwargs for anndata write function
+        """
+        # get all the public attributes
+        public_attributes = self._get_public_attributes()
+        # save the model state dict and the trainer state dict only
+        if not os.path.exists(dir_path) or overwrite:
+            os.makedirs(dir_path, exist_ok=overwrite)
+        else:
+            raise ValueError(
+                "{} already exists. Please provide an unexisting directory for saving.".format(
+                    dir_path
+                )
+            )
+
+        if save_anndata:
+            self.adata.write(
+                os.path.join(dir_path, "adata.h5ad"), **anndata_write_kwargs
+            )
+
+        model_save_path = os.path.join(dir_path, "model_params.pt")
+        attr_save_path = os.path.join(dir_path, "attr.pkl")
+        varnames_save_path = os.path.join(dir_path, "var_names.csv")
+
+        var_names = self.adata.var_names.astype(str)
+        var_names = var_names.to_numpy()
+        np.savetxt(varnames_save_path, var_names, fmt="%s")
+
+        torch.save(self.model.state_dict(), model_save_path)
+
+    def _load_expand_params_from_dict(self, state_dict):
```

Comment:
no point in having it at all, just delete it and replace with direct call of `load_state_dict`.